test/test_cuda.py

```diff
@@ -5613,6 +5613,123 @@ def my_function(pool):
             s = p.snapshot()
             self.assertEqual(len(s), 1, "Expected to have a single segment")
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    def test_graph_capture_reclaim_2_streams(self):
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:True"
+        )
+        torch.cuda.empty_cache()
+        s1 = torch.cuda.Stream()
+        s2 = torch.cuda.Stream()
+        g = torch.cuda.CUDAGraph(keep_graph=True)
+
+        torch.cuda.synchronize()
+
+        with torch.cuda.stream(s1):
+            g.capture_begin()
+            unrelated = torch.rand(8, device="cuda")
+
+            data1 = torch.rand(8, device="cuda")
+            data1_ptr = data1.data_ptr()
+
+            s2.wait_stream(s1)
+            with torch.cuda.stream(s2):
+                data1.fill_(2.0)
+                data1.record_stream(s2)
+
+            data1.fill_(1.0)
+            del data1
+
+            s1.wait_stream(s2)
+
+            # the graph would be joined after this kernel
+            unrelated.fill_(1.0)
+            # this new allocation should reuse data1 if the graph_capture_record_stream_reuse is enabled
+            data2 = torch.rand(8, device="cuda")
+            data2_ptr = data2.data_ptr()
+
+            g.capture_end()
+
+        torch.cuda.synchronize()
+        self.assertTrue(data1_ptr == data2_ptr)
+
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:False"
+        )
+
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs"
+    )
+    def test_graph_capture_reclaim_4_streams(self):
+        torch.cuda.memory._set_allocator_settings(
+            "graph_capture_record_stream_reuse:True"
+        )
+
+        torch.cuda.empty_cache()
+        s1 = torch.cuda.Stream()
+        s2 = torch.cuda.Stream()
+
+        s3 = torch.cuda.Stream()
+        s4 = torch.cuda.Stream()
+        g = torch.cuda.CUDAGraph(keep_graph=True)
+
+        torch.cuda.synchronize()
+
+        with torch.cuda.stream(s1):
+            g.capture_begin()
+            unrelated = torch.rand(8, device="cuda")
+
+            data1 = torch.ones(8, device="cuda")
+            data1_ptr = data1.data_ptr()
+
+            s2.wait_stream(s1)
+            with torch.cuda.stream(s2):
+                data1.fill_(2.0)
+                data1.record_stream(s2)
+
+            s3.wait_stream(s1)
+            with torch.cuda.stream(s3):
+                data1.fill_(3.0)
```

Comment:
from the correctness point of view, this script is confusing, you are filling data1 concurrently on s1, s2, s3 and s4